File: problems/967/solution.cpp
class Solution {
private:
    vector<int> ans;
    set<int> s;
    
public:
    vector<int> numsSameConsecDiff(int N, int K) {
        if (N == 1) {
            for (int i = 0; i < 10; i++) {
                ans.push_back(i);
            }
            
            return ans;
        }
        
        for (int i = 1; i < 10; i++) {
            dfs(i, N - 1, K);
        }
        
        ans.assign(s.begin(), s.end());
        return ans;
    }
    
    void dfs(int num, int n, int k) {
        if (n == 0) {
            s.insert(num);
            return;
        }
        
        int lastDigit = num % 10;
        
        if (lastDigit + k <= 9) {
            dfs(num * 10 + lastDigit + k, n - 1, k);
        }
        
        if (lastDigit - k >= 0) {
            dfs(num * 10 + lastDigit - k, n - 1, k);
        }
    }
};

File: problems/967/solution.py
class Solution:
    def getNum(self, ar):
        num = 0
        
        for n in ar:
            num *= 10
            num += n
        
        return num
    
    def numsSameConsecDiff(self, N: int, K: int) -> List[int]:
        if N == 1:
            return [i for i in range(10)]
        
        ans, q, i = set(), deque([[i] for i in range(1, 10)]), 1
        
        while q:
            cur = q.pop()
            
            if len(cur) > N:
                continue
            
            if len(cur) == N:
                ans.add(self.getNum(cur))
                continue
            
            j = cur[-1]
            
            if j + K <= 9: q.append(cur + [j + K])
            if j - K >= 0: q.append(cur + [j - K])
            # while j + K <= 9:
            #     q.append(cur + [j + K])
            #     j += K
            
            # j = cur[-1]
            # while j - K >= 0:
            #     q.append(cur + [j - K])
            #     j -= K
        
        return list(ans)